extensions/exquery/restxq/src/main/java/org/exist/extensions/exquery/restxq/impl/RestXqServiceImpl.java

```diff
@@ -85,23 +85,23 @@ public RestXqServiceImpl(final ResourceFunction resourceFunction, final BrokerPo
         this.brokerPool = brokerPool;
         this.binaryValueManager = new BinaryValueManager() {
 
-            final List<BinaryValue> binaryValues = new ArrayList<>();
+            final Deque<BinaryValue> binaryValues = new ArrayDeque<>();
 
             @Override
             public void registerBinaryValueInstance(final BinaryValue binaryValue) {
-                binaryValues.add(binaryValue);
+                binaryValues.push(binaryValue);
             }
 
             @Override
             public void runCleanupTasks() {
-                for (final BinaryValue binaryValue : binaryValues) {
+                while (!binaryValues.isEmpty()) {
                     try {
+                        final BinaryValue binaryValue = binaryValues.pop();
```

Comment:
in case if re-occuring IOexception this loop might never stop.

Reply:
It will stop when all items are popped